lib/spack/llnl/util/tty/log.py

```diff
@@ -105,8 +119,8 @@ def __exit__(self, exc_type, exception, traceback):
         """If termios was avaialble, restore old settings."""
         if self.old_cfg:
             import termios
-            termios.tcsetattr(
-                self.stream.fileno(), termios.TCSADRAIN, self.old_cfg)
+            with background_safe():  # change it back even if backgrounded now
```

Comment:
(question) It appears that `self.old_cfg` is only set if `not is_background()` in `__enter__`, so is this required?

Reply:
The process could've been moved to the foreground in between entering and exiting the context manager.

Reply:
I think I understand what you mean, but I only see an issue if it starts in the foreground and moves to the background in between (your statement implies the opposite).

It took some digging but https://docs.oracle.com/cd/E19455-01/806-0627/6j9vhfn8r/index.html appears to confirm that this is is a valid assumption (i.e. the config is actually set when SIGTTOU is ignored). If you know of a more-official source of that, IMO it would be worth adding the link as a comment.

Reply:
Yes, you are correct that it is the opposite case of what I said in my previous comment.

I do not know of a specific source, signal handling does not affect the behavior of the code throwing the signal, as a general rule.

Reply:
> signal handling does not affect the behavior of the code throwing the signal, as a general rule.

In the case that we aren't ignoring SIGTTOU, the system shouldn't apply the configuration to the tty: my understanding is that the reason SIGTTOU is raised is because the implementer assumes that background processes should not generally be able to change terminal settings that may affect foreground processes.

`tcsetattr` has to avoid applying the configuration in some cases but then I suppose it specifically checks if the signal is ignored and actually updates the config in that case.

Overall I think this bit of code is fine but it certainly wasn't obvious to me that it would work.